### ISSUE TITLE
supported_identity_providers in aws_cognito_user_pool_client Ordering Causing Differences

### ISSUE THREAD

**Issue by eangeeag** [terraform-providers/terraform-provider-aws#4806]
### Terraform Version

```bash
Terraform v0.11.7
+ provider.archive v1.0.0
+ provider.aws v1.22.0
```

### Affected Resource(s)

* ```aws_cognito_user_pool_client```

### Terraform Configuration Files

```hcl
resource "aws_cognito_user_pool" "kaizen_user_pool" {
  name                = "${var.project}-user-pool-${var.environment}"
  username_attributes = ["email"]

  schema {
    attribute_data_type = "String"
    name                = "email"
    required            = true

    # These are de defaults. If not specified, terraform would create/destroy this resource on 
    # every plan.
    string_attribute_constraints {
      min_length = 0
      max_length = 2048
    }
  }

  schema {
    attribute_data_type = "String"
    name                = "family_name"
    required            = true

    # These are de defaults. If not specified, terraform would create/destroy this resource on 
    # every plan.
    string_attribute_constraints {
      min_length = 0
      max_length = 2048
    }
  }

  schema {
    attribute_data_type = "String"
    name                = "name"
    required            = true

    # These are de defaults. If not specified, terraform would create/destroy this resource on 
    # every plan.
    string_attribute_constraints {
      min_length = 0
      max_length = 2048
    }
  }

  password_policy {
    minimum_length    = 8
    require_lowercase = true
    require_uppercase = true
    require_numbers   = true
    require_symbols   = true
  }

  admin_create_user_config {
    allow_admin_create_user_only = true
    unused_account_validity_days = 7
  }
}

resource "aws_cognito_identity_provider" "kaizen_saml" {
  user_pool_id  = "${aws_cognito_user_pool.kaizen_user_pool.id}"
  provider_name = "PeoplesTrustADFS"
  provider_type = "SAML"

  provider_details {
    MetadataURL = "<URL-here>"
  }

  attribute_mapping {
    name        = "<a-mapping>"
    family_name = "<a-mapping>"
    email       = "<a-mapping>"
  }
}

resource "aws_cognito_user_pool_client" "localhost" {
  name                   = "localhost"
  user_pool_id           = "${aws_cognito_user_pool.kaizen_user_pool.id}"
  refresh_token_validity = 30
  callback_urls          = ["https://localhost:3000/"]
  allowed_oauth_flows    = ["code", "implicit"]
  allowed_oauth_scopes   = ["phone", "email", "openid", "profile"]

  supported_identity_providers = ["${aws_cognito_identity_provider.kaizen_saml.provider_name}", "COGNITO"]
}
```

### Expected Behavior

Calling ` aws cognito-idp describe-user-pool-client` should return this:

```json
{
    "UserPoolClient": {
        "CallbackURLs": [
            "https://localhost:3000/"
        ],
        "AllowedOAuthScopes": [
            "email",
            "openid",
            "phone",
            "profile"
        ],
        "UserPoolId": "us-east-2_anId",
        "AllowedOAuthFlowsUserPoolClient": false,
        "LastModifiedDate": 1528729174.422,
        "ClientId": "another-id",
        "AllowedOAuthFlows": [
            "code",
            "implicit"
        ],
        "RefreshTokenValidity": 30,
        "CreationDate": 1528724981.009,
        "ClientName": "localhost"
    }
}
```

### Actual Behavior

Calling ` aws cognito-idp describe-user-pool-client` returns this, which is radically different:

```
{
    "UserPoolClient": {
        "UserPoolId": "us-east-2_EnBIWIdbg",
        "LastModifiedDate": 1528728467.294,
        "ClientId": "ldhmir6tqma88h3i7si53cau9",
        "AllowedOAuthFlowsUserPoolClient": false,
        "SupportedIdentityProviders": [
            "COGNITO",
            "PeoplesTrustADFS"
        ],
        "RefreshTokenValidity": 30,
        "CreationDate": 1528724981.009,
        "ClientName": "localhost"
    }
}
```

### Steps to Reproduce

1. `terraform plan`
2. `terraform apply`
3. `terraform plan`
4. `terraform apply`

EDIT: 
After further research it seems like the order of elements in `supported_identity_providers` changes on every/plan apply and this affects some other attributes of the `aws_cognito_user_poor_client` resource. 

Here's the `terraform plan` output of two successive `terraform plan`/`terraform apply` sequences without modifications to the code:

First time:
```
~ module.dev-cloud.module.cognito_user_pool.aws_cognito_user_pool_client.localhost
      allowed_oauth_flows.#:                  "0" => "2"
      allowed_oauth_flows.2645166319:         "" => "code"
      allowed_oauth_flows.3465961881:         "" => "implicit"
      allowed_oauth_scopes.#:                 "0" => "4"
      allowed_oauth_scopes.2517049750:        "" => "openid"
      allowed_oauth_scopes.2603607895:        "" => "phone"
      allowed_oauth_scopes.4080487570:        "" => "profile"
      allowed_oauth_scopes.881205744:         "" => "email"
      callback_urls.#:                        "0" => "1"
      callback_urls.0:                        "" => "https://localhost:3000/"
      supported_identity_providers.0:         "COGNITO" => "PeoplesTrustADFS"
      supported_identity_providers.1:         "PeoplesTrustADFS" => "COGNITO"
```

Second time:
```
 ~ module.dev-cloud.module.cognito_user_pool.aws_cognito_user_pool_client.localhost
      supported_identity_providers.0:         "COGNITO" => "PeoplesTrustADFS"
      supported_identity_providers.1:         "PeoplesTrustADFS" => "COGNITO"
```

EDIT #2:
Even though the second time I run `terraform plan`/`terraform apply` only shows me the change in the order of the `supported_identity_provider` elements:

```
 ~ module.dev-cloud.module.cognito_user_pool.aws_cognito_user_pool_client.localhost
      supported_identity_providers.0:         "COGNITO" => "PeoplesTrustADFS"
      supported_identity_providers.1:         "PeoplesTrustADFS" => "COGNITO"
```
Applying causes values of `allowed_oauth_flows` and `allowed_oauth_scopes` to be removed from the resource even though no indication of such behavior is indicated by the plan. If I run `terraform plan`/`terraform apply` again everything returns to normal.

**Comment by anouvel** [comment]
Hi @eangeeag,

I'm facing the same issue, but maybe I found a solution. After generating a skeleton (based on the AWS CLI), and compare it to the Terraform request, one field is missing:

```json
"ExplicitAuthFlows": [
    "CUSTOM_AUTH_FLOW_ONLY"
]
```

So basically, this one appears to work well:

```
resource "aws_cognito_user_pool_client" "client" {
  name = "my_name"

  refresh_token_validity               = 30
  generate_secret                      = false
  user_pool_id                         = "${aws_cognito_user_pool.pool.id}"
  allowed_oauth_flows                  = ["code", "implicit"]
  allowed_oauth_flows_user_pool_client = true
  allowed_oauth_scopes                 = ["aws.cognito.signin.user.admin", "phone", "email", "openid", "profile"]

  explicit_auth_flows          = ["CUSTOM_AUTH_FLOW_ONLY"]
  supported_identity_providers = ["COGNITO"]

  callback_urls = ["https://your_url"]

  read_attributes = [
    "address",
    "birthdate",
    "email",
    "email_verified",
    "family_name",
    "gender",
    "given_name",
    "locale",
    "middle_name",
    "name",
    "nickname",
    "phone_number",
    "phone_number_verified",
    "picture",
    "preferred_username",
    "profile",
    "updated_at",
    "website",
    "zoneinfo",
  ]

  write_attributes = [
    "address",
    "birthdate",
    "email",
    "family_name",
    "gender",
    "given_name",
    "locale",
    "middle_name",
    "name",
    "nickname",
    "phone_number",
    "picture",
    "preferred_username",
    "profile",
    "updated_at",
    "website",
    "zoneinfo",
  ]
}
```

**Comment by mmritesh** [comment]
I am also facing the similar issue.

**Comment by maanenh** [comment]
Same flip-flop here. Even with only one `supported_identity_providers`
Terraform v0.11.3
+ provider.archive v1.0.0
+ provider.aws v1.28.0

Is there a work-arround available?

**Comment by mmritesh** [comment]
@radeksimko Any update on the progress to solve this?

**Comment by bflad** [comment]
The fix for properly passing all attributes during `aws_cognito_user_pool_client` updates has been merged into master and will release with version 1.32.0 of the AWS provider, likely middle of next week. This should prevent the flip-flop apply behavior of the non-`supported_identity_providers` attributes.

We'll need to specifically address `supported_identity_providers` separately as it seems like the attribute might need to be `schema.TypeSet` instead of `schema.TypeList` from the descriptions above (I have not investigated anything personally though).

**Comment by bflad** [comment]
The followup fix for ignoring `supported_identity_providers` ordering has been merged and will release with version 2.54.0 of the Terraform AWS Provider, later this week. Thanks to @DrFaust92 for the implementation. 👍
